### DISCUSSION TITLE
Type errors when using ParamSpec in two functions

### DISCUSSION THREAD

**Discussion by Apakottur** [microsoft/pyright#7711] [Q&A] [answered]
Consider the following code:
```python
# test.py

from collections.abc import Callable
from typing import ParamSpec

P = ParamSpec("P")


class MyClass:
    def __init__(self, *, decorator: Callable[[Callable[P, None]], Callable[P, None]]) -> None:
        self._decorator = decorator

    def func(self, arg: Callable[P, None]) -> Callable[P, None]:
        return self._decorator(arg)
```

Running `pyright test.py` gives:
```text
test.py
  test.py:14:16 - error: Expression of type "(**P@__init__) -> None" cannot be assigned to return type "(**P@func) -> None"
    Type "(**P@__init__) -> None" cannot be assigned to type "(**P@func) -> None" (reportReturnType)
  test.py:14:32 - error: Argument of type "(**P@func) -> None" cannot be assigned to parameter of type "(**P@__init__) -> None"
    Type "(**P@func) -> None" cannot be assigned to type "(**P@__init__) -> None" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```

Is this a bug in Pyright or am I incorrectly using ParamSpec?
Mypy (strict) doesn't give any errors on this code.

**Comment by erictraut** [comment]
You have defined two methods that each use a separate function-scoped ParamSpec. I think you intended for the ParamSpec to be scoped to the class.

```python
class MyClass(Generic[P]): ...
```

Or if you're using Python 3.12, you can use the new syntax:

```python
class MyClass[**P]: ...
```

**Reply by Apakottur** [reply]
Thanks for the reply.
I did intend to use two separate function-scoped ParamSpec:
1. The first one is to make sure `decorator` doesn't change the signature of the function is receives.
2. The second one is to make sure `func` doesn't change the signature of the function is receivers.

If I use `Generic[P]`, I am then required to provide the type argument when creating `MyClass` objects which I don't want to do, `P` is arbitrary and the two functions are not necessarily used in tandem.

Why can't the function scoped `P` from the one be "put" into the other when used together?

These two functions are not necessarily even in the same class:
```python
# test.py
from collections.abc import Callable
from typing import ParamSpec

P = ParamSpec("P")


class MyClass:
    def __init__(self, *, decorator: Callable[[Callable[P, None]], Callable[P, None]]) -> None:
        self.decorator = decorator


# other_file.py
def func(c: MyClass, arg: Callable[P, None]) -> Callable[P, None]:
    return c.decorator(arg)
```

**Reply by erictraut** [reply, marked as answer]
What you're suggesting would make sense of `P` was scoped to the `Callable` type associated with the `decorator` parameter, but it's not. It's instead scoped to the `__init__` method, and it must be solved in that context. There's currently no way in the Python type system to specify that a type variable should be scoped to a `Callable`.

Both mypy and pyright include some heuristics that allow a function-scoped type variable to be "re-scoped" to a `Callable` type if that `Callable` is specified as a return type in a `def` statement. For example, `def func() -> Callable[[T], T]: ...`. But that is not specified in the typing spec and is arguably violating the spec [as it's currently worded](https://typing.readthedocs.io/en/latest/spec/generics.html#scoping-rules-for-type-variables).

This is an area that requires additional discussion and formal specification in the typing spec. We talked about it at some length when we were working on the design for [PEP 695](https://peps.python.org/pep-0695/), which introduced a new syntax for generics in Python 3.12. In particular, we looked at ways we could extend the syntax to allow type variables to be scoped to `Callable` type annotations rather than a `def` statement. We ultimately decided to punt on that because the PEP was already quite large, and adding more concepts to it would have made it less likely to be accepted.

If you're interested in exploring solutions to this in the context of the official typing spec, you're welcome to start a new discussion in the [typing forum](https://discuss.python.org/c/typing/32). This problem has been raised multiple times before, so there may be receptivity within the typing community to solving it.
